File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        int l = 0, r = 0, ans = 0;
        unordered_map<char, int> freq;

        while(r < s.size() && l < s.size() - 2) {
            freq[s[r]]++;
            if(freq['a'] > 0 && freq['b'] > 0 && freq['c'] > 0) {
                int rem = s.size() - r - 1;
                ans += rem + 1;
                freq[s[l]]--;
                freq[s[r]]--;
                l++;
            }
            else {
                r++;
            }
        }

        return ans;
    }
};

File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.ts
function numberOfSubstrings(s: string): number {
    let l = 0, freq = new Map<string, number>, ans = 0;
    for(let i = 0; i < s.length; i++) {
        freq.set(s[i], (freq.get(s[i]) ?? 0) + 1);

        while((freq.get('a') ?? 0) > 0 && (freq.get('b') ?? 0) > 0 && (freq.get('c') ?? 0) > 0) {
            ans += s.length - i;
            freq.set(s[l], freq.get(s[l]) - 1);
            l++;
        }
    }

    return ans;
};